python/ray/data/_internal/execution/operators/map_operator.py

```diff
@@ -75,6 +75,11 @@ def __init__(
         # All active `MetadataOpTask`s.
         self._metadata_tasks: Dict[int, MetadataOpTask] = {}
         self._next_metadata_task_idx = 0
+        # Keep track of all finished streaming generators.
+        # TODO(hchen): This is a workaround for a bug of lineage reconstruction.
+        # When the streaming generator ref is GC'ed, the objects it generated
+        # cannot be reconstructed. Should remove it once Ray Core fixes the bug.
+        self._finished_streaming_gens: List[StreamingObjectRefGenerator] = []
```

Comment:
if we keep the streaming generator ref here, does this prevent the underlying object refs to be GCed?

Reply:
do we need to cleanup `_finished_streaming_gens` in `all_inputs_done`?

Reply:
> if we keep the streaming generator ref here, does this prevent the underlying object refs to be GCed?
I don't think so. cc @rkooo567 to confirm.
> do we need to cleanup _finished_streaming_gens in all_inputs_done?
No, we need to keep them until the entire dataset finishes. Because even after all inputs are done, we may still need to re-run the tasks of this op.

Reply:
Can we clear this in .shutdown()?

Reply:
>Can we clear this in .shutdown()?

yeah, +1 to clear it in `shutdown()`

Reply:
added clean up in shutdown